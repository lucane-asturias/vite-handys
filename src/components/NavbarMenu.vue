<script setup>
  import { computed, reactive } from 'vue'
  import { useCartStore } from '@/store/cartStore'
  import { useProductStore } from '@/store/productStore'
  import formatPrice from '@/helpers/formatPrice'

  const productStore = useProductStore()
  const cartStore = useCartStore()

  const cartsLength = computed(() => cartStore.cartsLength)
  const cartsTotal = computed(() => cartStore.cartsTotal)
  
  const menu = reactive([
    { route: 'catalog', text: 'Products' },
    { route: 'cart', text: 'Cart' },
    { route: 'checkout', text: 'Order' }
  ])
</script>

<template>
  <header id="header" class="bg-slate-100 shadow drop-shadow-lg">
    <nav class="container mx-auto flex flex-row justify-start items-center p-8 px-2">
      <!-- Primary Navigation Menu -->
      <div v-for="item in menu" :key="item.route">
        <menu class="flex flex-row text-md mr-5">
          <router-link :to="{ name: item.route }" exact-active-class="text-blue-500">
            <span class="self-center text-md md:text-xl font-semibold" v-text="item.text" />
          </router-link>
        </menu>
      </div>
      <!-- Second Navigation Menu  -->
      <div class="flex flex-row ml-auto">
        <menu class="flex flex-row">
          <div class="mr-3">In Cart: {{ cartsLength }}</div>
          <div>Total: {{ formatPrice(cartsTotal) }}</div>
        </menu>
      </div>
    </nav>
  </header>
</template>