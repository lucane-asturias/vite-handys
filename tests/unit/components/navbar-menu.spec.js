import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import NavbarMenu from '@/components/NavbarMenu.vue'

describe('NavbarMenu.vue component', () => {

  let wrapper
  // SETUP - run prior to each unit test
  beforeEach(() => {
    setActivePinia(createPinia())

    wrapper = shallowMount(NavbarMenu, {
      global: { components: { 'router-link': RouterLinkStub }}
    })
  })

  test('should match snapshot', () => expect(wrapper.html()).toMatchSnapshot())

  test('should iterate three router-link menus', () => {
    expect(wrapper.findAllComponents(RouterLinkStub).length).toBe(3)
  })

  test.each`
    index | route         | expectedObj
    ${0}  | ${'catalog'}  | ${{ name: 'catalog' }}
    ${1}  | ${'cart'}     | ${{ name: 'cart' }}
    ${2}  | ${'checkout'} | ${{ name: 'checkout' }}
  `(
    'menu items should navigate to their respective route: $route', 
    ({ index, expectedObj }) => {
      expect(wrapper.findAllComponents(RouterLinkStub).at(index).props().to)
      .toMatchObject(expectedObj)
    }
  )
  
  test('default values for computed properties should be 0', () => {
    expect(wrapper.vm.cartsTotal).toBe(0)
    expect(wrapper.vm.cartsLength).toBe(0)
    expect(wrapper.text()).toContain('In Cart: 0')
    expect(wrapper.text()).toContain('Total: $0.00')
  })

})