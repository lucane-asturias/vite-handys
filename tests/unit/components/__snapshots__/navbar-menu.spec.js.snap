// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavbarMenu.vue component > should match snapshot 1`] = `
"<header id=\\"header\\" class=\\"bg-slate-100 shadow drop-shadow-lg\\">
  <nav class=\\"container mx-auto flex flex-row justify-start items-center p-8 px-2\\">
    <!-- Primary Navigation Menu -->
    <div>
      <menu class=\\"flex flex-row text-md mr-5\\">
        <router-link-stub to=\\"[object Object]\\" custom=\\"false\\" exact-active-class=\\"text-blue-500\\"></router-link-stub>
      </menu>
    </div>
    <div>
      <menu class=\\"flex flex-row text-md mr-5\\">
        <router-link-stub to=\\"[object Object]\\" custom=\\"false\\" exact-active-class=\\"text-blue-500\\"></router-link-stub>
      </menu>
    </div>
    <div>
      <menu class=\\"flex flex-row text-md mr-5\\">
        <router-link-stub to=\\"[object Object]\\" custom=\\"false\\" exact-active-class=\\"text-blue-500\\"></router-link-stub>
      </menu>
    </div><!-- Second Navigation Menu  -->
    <div class=\\"ml-auto\\">
      <menu class=\\"flex flex-row\\">
        <div class=\\"mr-3\\">In Cart: 0</div>
        <div>Total: $0.00</div>
      </menu>
    </div>
  </nav>
</header>"
`;
